File: eval/src/retrieval/bm25.py
from __future__ import annotations

import math
import re
from collections import Counter

from src.retrieval.types import SearchResult


class BM25Index:
    def __init__(self, k1: float = 1.5, b: float = 0.75):
        self.k1 = k1
        self.b = b
        self._chunks: list[dict] = []
        self._term_counts: list[Counter[str]] = []
        self._doc_freqs: Counter[str] = Counter()
        self._avg_doc_len = 0.0

    def index_chunks(self, chunks: list[dict]) -> None:
        self._chunks = list(chunks)
        self._term_counts = []
        self._doc_freqs = Counter()

        total_terms = 0
        for chunk in self._chunks:
            counts = Counter(_tokenize(chunk["text"]))
            self._term_counts.append(counts)
            self._doc_freqs.update(counts.keys())
            total_terms += sum(counts.values())

        self._avg_doc_len = total_terms / max(1, len(self._term_counts))

    def search(self, query: str, top_k: int) -> list[SearchResult]:
        query_terms = _tokenize(query)
        if not query_terms or not self._chunks:
            return []

        scored: list[SearchResult] = []
        for chunk, counts in zip(self._chunks, self._term_counts, strict=True):
            score = self._score(query_terms, counts)
            if score <= 0:
                continue
            scored.append(
                SearchResult(
                    chunk_id=chunk["id"],
                    doc_id=chunk["doc_id"],
                    text=chunk["text"],
                    score=score,
                    source=chunk.get("source", ""),
                )
            )

        scored.sort(key=lambda hit: hit.score, reverse=True)
        return scored[:top_k]

    def _score(self, query_terms: list[str], counts: Counter[str]) -> float:
        score = 0.0
        doc_len = sum(counts.values())
        for term in query_terms:
            freq = counts.get(term, 0)
            if freq == 0:
                continue

            idf = self._idf(term)
            denom = freq + self.k1 * (1 - self.b + self.b * doc_len / max(1.0, self._avg_doc_len))
            score += idf * (freq * (self.k1 + 1)) / denom
        return score

    def _idf(self, term: str) -> float:
        n_docs = len(self._chunks)
        doc_freq = self._doc_freqs.get(term, 0)
        return math.log(1 + (n_docs - doc_freq + 0.5) / (doc_freq + 0.5))


def _tokenize(text: str) -> list[str]:
    return re.findall(r"[a-z0-9]+", text.lower())

File: web/src/lib/server/bm25.ts
import type { Chunk, SearchHit } from "./types";

function tokenize(text: string): string[] {
  return (text.toLowerCase().match(/[a-z0-9]+/g) || []);
}

export class BM25Index {
  private k1 = 1.5;
  private b = 0.75;
  private chunks: Chunk[] = [];
  private termCounts: Map<string, number>[] = [];
  private docFreqs = new Map<string, number>();
  private avgDocLen = 0;

  indexChunks(chunks: Chunk[]): void {
    this.chunks = [...chunks];
    this.termCounts = [];
    this.docFreqs = new Map();
    let totalTerms = 0;

    for (const chunk of this.chunks) {
      const counts = new Map<string, number>();
      for (const term of tokenize(chunk.text)) {
        counts.set(term, (counts.get(term) || 0) + 1);
      }
      this.termCounts.push(counts);
      for (const term of counts.keys()) {
        this.docFreqs.set(term, (this.docFreqs.get(term) || 0) + 1);
      }
      for (const n of counts.values()) totalTerms += n;
    }
    this.avgDocLen = totalTerms / Math.max(1, this.termCounts.length);
  }

  /**
   * Score every chunk with a positive BM25 contribution.
   * Callers use a large fetch_k then diversity-select (MMR / multi-doc).
   */
  scoreAll(query: string): SearchHit[] {
    const queryTerms = tokenize(query);
    if (!queryTerms.length || !this.chunks.length) return [];

    const scored: SearchHit[] = [];
    for (let i = 0; i < this.chunks.length; i++) {
      const chunk = this.chunks[i];
      const counts = this.termCounts[i];
      const score = this.score(queryTerms, counts);
      if (score <= 0) continue;
      scored.push({
        chunk_id: chunk.id,
        doc_id: chunk.doc_id,
        source: chunk.source,
        text: chunk.text,
        score,
      });
    }
    scored.sort((a, b) => b.score - a.score);
    return scored;
  }

  search(query: string, topK: number): SearchHit[] {
    return this.scoreAll(query).slice(0, topK);
  }

  private score(queryTerms: string[], counts: Map<string, number>): number {
    let score = 0;
    let docLen = 0;
    for (const n of counts.values()) docLen += n;
    for (const term of queryTerms) {
      const freq = counts.get(term) || 0;
      if (!freq) continue;
      const idf = this.idf(term);
      const denom = freq + this.k1 * (1 - this.b + (this.b * docLen) / Math.max(1, this.avgDocLen));
      score += (idf * (freq * (this.k1 + 1))) / denom;
    }
    return score;
  }

  private idf(term: string): number {
    const nDocs = this.chunks.length;
    const docFreq = this.docFreqs.get(term) || 0;
    return Math.log(1 + (nDocs - docFreq + 0.5) / (docFreq + 0.5));
  }
}
